lc-hax/Scripts/Helpers/Enemies.cs

```diff
@@ -19,6 +20,61 @@ internal static partial class Helper {
         enemyInstance.SwitchToBehaviourServerRpc(Convert.ToInt32(state));
     }
 
+    internal static PlayerControllerB? GetPlayerAboutToKilledByEnemy(int playerObjectID) {
+        PlayerControllerB[] players = Helper.Players;
+        return players.First(player => (int)player.playerClientId == playerObjectID);
+    }
+
+    internal static bool IsLocalPlayerAboutToGetKilledByEnemy(int PlayerID) {
+        PlayerControllerB? player = Helper.GetPlayerAboutToKilledByEnemy(PlayerID);
+        return player != null && player.IsSelf();
+    }
+
+    internal static bool IsLocalPlayerAboutToGetKilledByEnemy(this EnemyAI instance, Collider other) {
+        if (instance == null) return false;
+        if (other == null) return false;
+        PlayerControllerB playerControllerB = instance.MeetsStandardPlayerCollisionConditions(other, false, false);
+        return playerControllerB != null && playerControllerB.IsSelf();
+    }
+
+    private static Dictionary<string, GameObject>? _AllSpawnableEnemies;
+
+    internal static Dictionary<string, GameObject> AllSpawnableEnemies {
+        get {
+            if (_AllSpawnableEnemies == null) {
+                _AllSpawnableEnemies = [];
+
+                HashSet<(string Name, GameObject Prefab)> uniqueEnemies = [];
+
+                SelectableLevel[] levels = Resources.FindObjectsOfTypeAll<SelectableLevel>();
+
+                foreach (SelectableLevel level in levels) {
+                    List<List<SpawnableEnemyWithRarity>> enemyCollections = [
+                        level.Enemies,
+                        level.OutsideEnemies,
+                        level.DaytimeEnemies
+                    ];
+
+                    foreach (List<SpawnableEnemyWithRarity> collection in enemyCollections) {
+                        foreach (SpawnableEnemyWithRarity enemy in collection) {
+                            if (enemy.enemyType.enemyName.Contains("Docile Locust Bees", StringComparison.OrdinalIgnoreCase)) continue;
+                            if (enemy.enemyType.enemyName.Contains("Manticoil", StringComparison.OrdinalIgnoreCase)) continue;
+                            _ = uniqueEnemies.Add((enemy.enemyType.enemyName, enemy.enemyType.enemyPrefab));
+                        }
+                    }
+                }
+
+                foreach ((string Name, GameObject Prefab) in uniqueEnemies) {
+                    _ = _AllSpawnableEnemies.TryAdd(Name, Prefab);
+                }
+            }
+
+            return _AllSpawnableEnemies;
+        }
+    }
```

Comment:
What's the point of this? Where will we be reusing this?

Reply:
is being used in /spawn and /enemies.
it does it once as it parses all prefabs and names from all loaded levels , including modded ones and modded enemies.

Reply:
> that was the last one, im enchancing /spawn to cover all enemies of all levels, not just the ones based off that level, 

Didn't you mention this? If you could spawn all enemies across levels, why do we need to check for this?

Reply:
Oh, that's because each SelectableLevel has the prefabs registered there, i dont know any other way to parse it.

Reply:
```cs
Resources.FindObjectsOfTypeAll<EnemyType>().Select(enemyType => enemyType.enemyPrefab);
```

Reply:
Fixed

Reply:
I give up.

Reply:
> I give up.

you really want to rely always on Resources.FindObjectsOfTypeAll and not on a filled dictionary that will fill once?

Reply:
Nope. I am more amazed that you would rather use 2 for-loops instead of using `Where` and `Select`.

Reply:
> Nope. I am more amazed that you would rather use 2 for-loops instead of using `Where` and `Select`.

??? is just one foreach!

Reply:
You're right, but you could just return directly if you did it with method chaining rather than this mess.

```cs

            if (_AllSpawnableEnemies == null) {
                _AllSpawnableEnemies = [];
                foreach (EnemyType enemy in Resources.FindObjectsOfTypeAll<EnemyType>()) {
                    if (enemy.enemyName.Contains("Docile Locust Bees", StringComparison.OrdinalIgnoreCase)) continue;
                    if (enemy.enemyName.Contains("Manticoil", StringComparison.OrdinalIgnoreCase)) continue;
                    _ = _AllSpawnableEnemies.TryAdd(enemy.enemyName, enemy.enemyPrefab);
                }
            }

            return _AllSpawnableEnemies;

```